akka-actor-typed/src/main/scala/akka/actor/typed/internal/pubsub/TopicImpl.scala

```diff
@@ -0,0 +1,129 @@
+@InternalApi
+private[akka] object TopicImpl {
+
+  trait Command[T]
+
+  // actual public messages but internal to ease bincomp evolution
+  final case class Publish[T](message: T) extends Topic.Command[T]
+  final case class Subscribe[T](subscriber: ActorRef[T]) extends Topic.Command[T]
+  final case class Unsubscribe[T](subscriber: ActorRef[T]) extends Topic.Command[T]
+
+  final case class TopicInstancesUpdated[T](topics: Set[ActorRef[TopicImpl.Command[T]]]) extends Command[T]
+  final case class MessagePublished[T](message: T) extends Command[T]
+  final case class SubscriberDied[T](subscriber: ActorRef[T]) extends Command[T]
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[akka] final class TopicImpl[T](topicName: String, context: ActorContext[TopicImpl.Command[T]])(
+    implicit classTag: ClassTag[T])
+    extends AbstractBehavior[TopicImpl.Command[T]](context) {
+
+  /*
+   * The topic actor keeps a local set of subscribers, whenever that is non-empty it registers itself for
+   * a topic service key, and when it becomes empty again it deregisters from the service key. Published
+   * messages go to all currently known topics registered for the topic service key and the individual topic
+   * instances then broadcast the message to all local subscribers. This achieves deduplication for nodes
+   * with multiple subscribers and avoids sending to nodes without any subscribers.
+   */
+
+  import TopicImpl._
+
+  private val topicServiceKey = ServiceKey[TopicImpl.Command[T]](topicName)
+  context.log.debugN("Starting up pub-sub topic [{}] for messages of type [{}]", topicName, classTag.runtimeClass)
+
+  private var topics = Set.empty[ActorRef[TopicImpl.Command[T]]]
+  private var localSubscribers = Set.empty[ActorRef[T]]
+
+  private def receptionist = context.system.receptionist
+  private val receptionistAdapter = context.messageAdapter[Receptionist.Listing] {
+    case topicServiceKey.Listing(topics) => TopicInstancesUpdated(topics)
+  }
+  receptionist ! Receptionist.Subscribe(topicServiceKey, receptionistAdapter)
+
+  override def onMessage(msg: TopicImpl.Command[T]): Behavior[TopicImpl.Command[T]] = msg match {
+
+    case Publish(message) =>
+      if (topics.isEmpty) {
+        context.log.trace("Publishing message of type [{}] but no subscribers, dropping", msg.getClass)
+        context.system.deadLetters ! Dropped(message, "No topic subscribers known", context.self.toClassic)
+      } else {
+        context.log.trace("Publishing message of type [{}]", msg.getClass)
+        val pub = MessagePublished(message)
+        topics.foreach(_ ! pub)
+      }
+      this
+
+    case MessagePublished(msg) =>
+      context.log.trace("Message of type [{}] published", msg.getClass)
+      localSubscribers.foreach(_ ! msg)
+      this
+
+    case Subscribe(subscriber) =>
+      context.watchWith(subscriber, SubscriberDied(subscriber))
+      localSubscribers = localSubscribers + subscriber
+      if (localSubscribers.size == 1) {
```

Comment:
if the same actor double registers we'll subscribe multiple times to the receptionist, maybe that is okay

Reply:
Let's check and ignore for that case